### PR TITLE
AVRO-3917: [Rust] take field aliases into account when calculating schema compatibilities

### PR DESCRIPTION
**Pull request by marcosschroh**
## What is the purpose of the change

- Take field aliases are not taken into account when calculating schema compatibility [AVRO-3917](https://issues.apache.org/jira/browse/AVRO-3917)

## Verifying this change

This change is partially covered by existing tests, such as `avro_3894_take_aliases_into_account_when_serializing_for_schema_compatibility`,  however it does not cover all the `aliases` use cases in a proper way. With the changes that I am adding then it will fix the issue and we will covered the missing uses cases. The new test added is take_aliases_into_account_for_schema_compatibility

## Documentation

- Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: lang/rust/avro/src/schema_compatibility.rs [hunk 1]
```diff
@@ -1183,23 +1213,38 @@ mod tests {
...
-        pub struct Conference {
-            pub name: String,
-            pub date: i64,
-        }
-        #[derive(Debug, PartialEq, Eq, Clone, Deserialize, Serialize)]
-        pub struct ConferenceV2 {
-            pub name: String,
-            pub time: i64,
-        }
+        let schema_v3 = Schema::parse_str(
```
Comment (martin-g) [inline lang/rust/avro/src/schema_compatibility.rs:1219]: Please create new test methods instead of extending existing ones.
It is hard to follow what exactly is being tested when the test tests more than one use case.

File: lang/rust/avro/src/schema_compatibility.rs [hunk 2]
```diff
@@ -158,17 +158,45 @@ impl Checker {
...
-                    if let Some(pos) = w_lookup.get(&field.name) {
-                        if let Err(err) =
-                            self.full_match_schemas(&w_fields[*pos].schema, &field.schema)
-                        {
-                            return Err(CompatibilityError::FieldTypeMismatch(
-                                field.name.clone(),
-                                Box::new(err),
-                            ));
+                    // get all field names in a vector (field.name + aliases)
+                    let mut fields_names = vec![field.name.clone()];
```
Comment (martin-g) [inline lang/rust/avro/src/schema_compatibility.rs:162]: I think you can use a `Vec<&str>` instead, and avoid the clones.

**Review by martin-g — CHANGES_REQUESTED** [review]

File: lang/rust/avro/src/schema_compatibility.rs (hunk 2 above)
Comment (marcosschroh) [reply lang/rust/avro/src/schema_compatibility.rs:162]: Cool!! Thanks

**Comment by marcosschroh** [comment]
@martin-g I am seeing that the CI is failing because some packages can not be installed as they `required rustc 1.70.0 or newer`, for example `clap builder` and `anstyle`. Shall we update the `rust` version to ` 1.70.0`?

**Review by martin-g — APPROVED** [review]

**Comment by martin-g** [comment]
Thank you, @marcosschroh !

File: lang/rust/avro/src/schema_compatibility.rs [hunk 3]
```diff
@@ -158,17 +158,45 @@ impl Checker {
...
+                    if let Some(ref aliases) = field.aliases {
+                        for alias in aliases {
+                            fields_names.push(alias);
+                        }
+                    }
+
+                    // Check whether any of the possible fields names are in the writer schema.
+                    // If the field was found, then it must have the exact same name with the writer field,
+                    // otherwise we would have a false positive with the writers aliases
+                    let mut position = None;
```
Comment (woile) [inline lang/rust/avro/src/schema_compatibility.rs:172]: you can avoid this `let mut` with a find:
```rs
let position = fields_names.iter().find(|field_name| {
    if let Some(pos) = w_lookup.get(field_name) {
        if &w_fields[*pos].name == field_name {
            return Some(pos);
        }
    }
    return None
})
```
